### ISSUE TITLE
Cloudstack Management Improper shutdown - causes mgmt to not start after reboot of server

### ISSUE THREAD

**Issue by rvalle** [apache/cloudstack#3954]
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Cloudstack Manager
~~~

##### CLOUDSTACK VERSION

~~~
4.13.0.0
~~~

##### OS / ENVIRONMENT

Centos 7, following quick installation guide.

##### SUMMARY

I believe a Cloudstack Management Server shutdown will shutdown mysql and the Cloudstack Manager in no particular order, which results in instability.

##### STEPS TO REPRODUCE

~~~
Install the Cloudstack Management server followign the quick install guide,
when finished, reboot the system.
A lot of exceptions appear in the log files during the Manager shutdown pointing at the database not being available.

The system may become "broken" after restart.

~~~

##### EXPECTED RESULTS

~~~
The server should be rebooted cleanly, all components (manager, mysql, etc) in the right order, no exceptions thrown.
~~~

##### ACTUAL RESULTS

~~~
Lots of exceptions during shutdown.
Non working system after reboot.



~~~

**Comment by rvalle** [comment]
I am pretty sure the installed Systemd scripts don't observe the order in which components should be shutdown, can anybody confirm?

I will get 2 pastebins with the exceptions during the shutdown and after restart.

**Comment by rvalle** [comment]
Exceptions during shutdown: (to be provided)
Exceptions after restart: https://pastebin.com/RWRPCSd1

**Comment by DaanHoogland** [comment]
@rvalle can you confirm this with 4.14 as well?

**Comment by rvalle** [comment]
@DaanHoogland should be feasible. are there packages in any centos repo?

**Comment by rvalle** [comment]
I cannot find 4.14 in the release repo. Is there a nightly build repo? If there is a repo I can run my tests easily. And could also check 4.11 LTS

**Comment by DaanHoogland** [comment]
@rvalle 4.14 is not released and would be a snapshot of master. if you can't build yourself there are nightly builds around.

**Comment by rvalle** [comment]
@DaanHoogland I am currently migrating/extending an ansible collection of roles to manage my cloud form opennebula to cloudstack. The interesting thing is that I did setup a cloud Continuous Integration infrastructure. On every push a 3 node cloudstack cluster is created and tested .
 
Let me advance a little bit more on this front and I would be happy to integrate this CI with a Nightly Builds or Travis that generates packages that I can take. We can discuss.

![Screenshot_2020-03-10 Pipeline · privazio ansible-collection-vdc](https://user-images.githubusercontent.com/412837/76330715-7c937b00-62ee-11ea-97cd-6b52146ac9c7.png)

**Comment by rvalle** [comment]
Systemd seems to support 2 approaches,
- The obvious one, to fix the provided Systemd files for CS Manager.
- The second one, is to supplement existing configuration files with additional files that fine tune its behaviour.
The second one is useful in the (unlikely) case we needed to modify mysql unit file or in case we want to patch installation based on versions of CS Manager with this bug.

**Comment by DaanHoogland** [comment]
I think the best solution will be somewhere in the middle. @rvalle, are you taking the time to propose a change?

**Comment by rvalle** [comment]
@DaanHoogland I can certainly put together a PR for option 1, to add the dependency to Mysql the mysql service.

As you tagged the issue I believe it is critical. In particular with a reboot right after installation. By the log files, I believe you are using a Java Persistence framework, and probably it is likely that will have a cache layer. 

I still think that it is useful do document the second approach somewhere, becuse it is likely that any deployment will have additional dependencies.

**Comment by rvalle** [comment]
For option 2, in case you need to supplement your existing system with an additional dependency just:
- create the directory: **/etc/systemd/system/cloudstack-management.service.d**
- create a configuration file, such as **mysq-dep.conf** for that particular dependency in that directory, with the following content:
```
[Unit]
Requires=mysql.service
After=mysql.service
```
- then you need to reload systemd configuration with **systemctl daemon-reload**
- test that the dependency is working: for stopping mysql will stop your cloudstack management service first.
- Similar files can be created to add other dependencies: ldap, nginx, etc. depending on your setup.

**Comment by rvalle** [comment]
@DaanHoogland the problem with option 1 is that we don't really know if the mysql server is running at the management service host or not. 

According to documentation the database server can also run on a separate server.

Perhaps we should use option 2 from cloudstack-setup-databases

**Comment by DaanHoogland** [comment]
@rvalle sound good, one doubt is editting the unit dep file with ldap and/or nginx. How will that behave on upgrades? for now ldap is only used readonly so that would not be a big problem.

**Comment by DaanHoogland** [comment]
btw, I could read above that you think it should be solved in documentation and not in packaging. Is that correct?

**Comment by rvalle** [comment]
@DaanHoogland 
when supplementing a service profile it is possible to create a directory under /etc/systemd... (as shown avobe) and under the directory an independent file can be created to map each dependency.

systemd foresees this particular use case where dependencies need to be tuned by the system administrator, independently of the software distribution.

Package distributions will install their service units under /lib/systemd and will not interfere.

In other words, I can add an ldap-dep.conf file and an nginx-dep.conf in /etc/systemd/system/cloudstack-management.service.d/ and they will complement the file provided by the distribution, even if there are updates (directory trees are independent).

Having said this, we face 2 problems.
- Admins may or may not run mysql alongside the management server
- Admins may or may not run other services (ldap, nginx, etc).

What I think we could implement is:
- leave **/lib/systemd/system/cloudstack-management.service** unchanged, because we don't know where mysql will run, and we cannot ask the admins to patch that file according to their env.
- modify **cloudstack-setup-databases** to create a service profile supplement by default, creating the dependency as: **/etc/systemd/system/cloudstack-management.service.d/mysql-dep.conf** with an option to disable its creation for the cases in which both servers won't run alongside.
- include a reference in the documentation. perhaps a link to the redhat configuration above, also in form of comments on the distributed service file and the file created by **cloudstack-setup-databases**

**Comment by andrijapanicsb** [comment]
@rvalle I'm trying to reproduce this on the current mater.
In the meantime, what do you mean that the system is broken after the restart? Can you describe that? thx

**Comment by andrijapanicsb** [comment]
I guess I answered this question myslef by testing - the mgmt will not boot after the MySQL and thus is down, needs restart of mgmt server to be up again.

Marking this one as blocker against 4.14, but the fix will be implemented un 4.13.1 so we can forwar merge to 4.14/master.

Thanks @rvalle for reporting this, nice blocker :)

**Comment by rvalle** [comment]
@andrijapanicsb  I found out about this in issue #3909 a reboot after the initial setup is actually fatal.
you are welcome, happy to contribute when I can.

**Comment by rhtyd** [comment]
@rvalle on reboot could it be that mysql is shutdown before Cloudstack management server? You can check using systemd, or just add a dependency on mysql service to shutdown after cloudstack mgmt server.

**Comment by rvalle** [comment]
@rhtyd yes, that is the case. However, CSM has a long starting time while mysql is quick. In a race condition I think mysql wins. I think corruption takes place on shutdown depending on what the DB was up to. 

In my use case I deploy CSM with ansible and then restart. I guess the DB was pretty busy at shutdown time to the point that every single test would fail.

I have been testing the dependencies introduced in /etc/systemd listed above, and they work like a charm.

**Comment by DaanHoogland** [comment]
@rvalle How do you want to proceed this?
I can imagine a wait in the automation scripts could help as well as a dependency in the systemd configuration. That last one however could lead to a system fault as the dependency does not have to exist locally on the host. Have you given this any thought over the weekend?
I am open to any solution if it works with your automation (and doesn't destroy the rest of the world).

**Comment by DaanHoogland** [comment]
@rvalle @rhtyd @andrijapanicsb I've read up on systemd a bit;

1. `If unit1 has the directive After=unit2, then if both units are run, unit2 will be executed fully before unit1 starts.`

1. `If unit1 has Wants=unit2 as a dependency, when unit1 is run, unit2 will be run as well. But whether unit2 starts successfully does not affect unit1 running successfully.`

I think we can leverage the two to make sure that if mysql is installed on the same server it will be properly schedule to start before and shutdown after cloudstack.

**Comment by DaanHoogland** [comment]
issue with that last solution is that the target to depend on may be different:
it may be mariadb or mysql or mysqld depnding on distribution used.
maybe solve is docs after all :(

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
fixed by #3973
